### PR TITLE
[FSDP2] Added 2D DCP save/load test

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120952
* #121680
* #121816
* __->__ #121747

To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load. For loading into a new model/optimizer instance, torchtrain uses lightweight `ModelWrapper` and `OptimizerWrapper`. In the added unit test, we use `get_optimizer_state_dict` directly to show the minimal requirement for correctness.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 1]
```diff
@@ -393,7 +393,13 @@ def _get_grad_inner_tensor(self, grad: torch.Tensor) -> torch.Tensor:
         if self.is_dtensor:
             if isinstance(grad, AsyncCollectiveTensor):
                 grad = grad.wait()
-            grad = cast(DTensor, grad)._local_tensor
+            assert isinstance(grad, DTensor), f"{type(grad)}"
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:396]: Include this fix for partial gradients (e.g. for replicated `norm.weight`)

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 2]
```diff
@@ -710,13 +712,16 @@ class TestFullyShard2DTraining(FSDPTest):
     def world_size(self) -> int:
         return min(4, torch.cuda.device_count())
 
-    @skip_if_lt_x_gpu(2)
-    def test_train_parity_2d_mlp(self):
+    def init_global_mesh(self) -> DeviceMesh:
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:715]: just factor this global mesh init into a method

**Comment by awgu** [comment]
Convert to draft since I will look at loading into new model/optimizer instance and maybe include that in this PR too

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 3]
```diff
@@ -710,13 +713,16 @@ class TestFullyShard2DTraining(FSDPTest):
...
         # Prefer to test with >=4 GPUs, but for 2 GPUs, use 2-way TP
         dp_size = 2 if self.world_size > 2 else 1
-        global_mesh = init_device_mesh(
+        return init_device_mesh(
             "cuda", (dp_size, self.world_size // dp_size), mesh_dim_names=("dp", "tp")
         )
+
+    @skip_if_lt_x_gpu(2)
+    def test_train_parity_2d_mlp(self):
```
Comment (wz337) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:724]: Should we make optimizer an option in .run_subtests? 
Actually, I can try this after you land, since I have observed parameters are off by a little bit more than what torch.allclose() tolerates by default in the unit tests.

Context: https://github.com/pytorch/pytorch/pull/121774
Local test is fine but CI is failing.

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 4]
```diff
@@ -782,6 +788,106 @@ def _test_train_parity_2d_mlp(
...
+        # model/optimizer, load checkpoint, and run another iteration
+        torch.manual_seed(seed)
+        model_cp = parallelize(Transformer(model_args), global_mesh, use_seq_parallel)
+        optim_cp = torch.optim.Adam(model_cp.parameters(), lr=1e-2)
+
+        loss_cp1 = train_step(model_cp, optim_cp, inp)
+        self.assertEqual(loss_no_cp1, loss_cp1)
+
+        sharded_sd = {
+            "model": model_cp.state_dict(),
```
Comment (wz337) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:852]: For the sake of testing DCP checkpoint, could we also use `get_model_state_dict()`, as we expect `model.state_dict()` to be identical as `get_model_state_dict()`?

**Review by wz337 — APPROVED** [review]
LGTM!

The comments are minor. If it is easier for you, you can land it as it is. I can try AdamW and update model state dict to use distributed state dict as well.

File: test/distributed/_composable/fsdp/test_fully_shard_training.py (hunk 3 above)
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:727]: I think since Adam is fine/passing, maybe we can leave the test always testing optimizer here since we want to check the end-to-end resume flow? Then, you can investigate the AdamW case separately?
